### DISCUSSION TITLE
SetDirectory not honored after WiX v4 upgrade?

### DISCUSSION THREAD

**Discussion by tristanlabelle** [wixtoolset/issues#7330] [Questions]
Hello,

I'm porting the installer for the Swift programming language toolchain for Windows to WiX v4. I ran the converter from WiX v3 and the resulting installer doesn't seem to honor the `<SetDirectory>` that is supposed to set the default installation location anymore. The result is that the files get installed to `C:\...` instead of `C:\Program Files\swift\...`. What am I missing?

See `platforms/Windows/runtime-amd64.wxs` in: https://github.com/apple/swift-installer-scripts/commit/2a30cdec11e07516254ae1e30d1a733be6b51add

**Comment by robmen** [comment]
How does the verbose log file for the two versions differ? IIRC (off the top of my head), there was a fix to more correctly schedule `SetDirectory` that may cause your `Condition` to be evaluated differently.

PS: I could be missing something by the design of some of your directory trees looks overly complex. You don't need `SetDirectory` to install to `ProgramFilesFolder`. And there seems to be lots of duplicated code.

**Reply by tristanlabelle** [reply]
I notice that when I run the v4 installer, the installation path textbox in the UI is set to `C:\` by default instead of `C:\Program Files\swift` as was the case with v3.

I compared the log files using `msiexec /i "runtime.msi" LOGVERBOSE=1 /l*v install.log` and it looks like the execution of `SetINSTALLDIR` is reordered to happen too late, so it gets skipped?

v3:
```
MSI (c)  []: Doing action: SetINSTALLDIR
MSI (c)  []: Note: 1: 2205 2:  3: ActionText 
Action  SetINSTALLDIR. 
Action start  SetINSTALLDIR.
MSI (c)  []: PROPERTY CHANGE: Adding INSTALLDIR property. Its value is 'C:\Program Files\swift'.
Action ended  SetINSTALLDIR. Return value 1.
...
MSI (c)  []: PROPERTY CHANGE: Adding TARGETDIR property. Its value is 'C:\'.
MSI (c)  []: PROPERTY CHANGE: Modifying INSTALLDIR property. Its current value is 'C:\Program Files\swift'. Its new value: 'C:\Program Files\swift\'.
MSI (c)  []: PROPERTY CHANGE: Adding _ property. Its value is 'C:\Program Files\swift\runtime-development\'.
MSI (c)  []: PROPERTY CHANGE: Adding _usr property. Its value is 'C:\Program Files\swift\runtime-development\usr\'.
MSI (c)  []: PROPERTY CHANGE: Adding _usr_bin property. Its value is 'C:\Program Files\swift\runtime-development\usr\bin\'.
MSI (c)  []: Target path resolution complete. Dumping Directory table...
MSI (c)  []: Note: target paths subject to change (via custom actions or browsing)
MSI (c)  []: Dir (target): Key: TARGETDIR	, Object: C:\
MSI (c)  []: Dir (target): Key: INSTALLDIR	, Object: C:\Program Files\swift\
MSI (c)  []: Dir (target): Key: _	, Object: C:\Program Files\swift\runtime-development\
MSI (c)  []: Dir (target): Key: _usr	, Object: C:\Program Files\swift\runtime-development\usr\
MSI (c)  []: Dir (target): Key: _usr_bin	, Object: C:\Program Files\swift\runtime-development\usr\bin
```

v4:
```
MSI (c)  []: PROPERTY CHANGE: Adding TARGETDIR property. Its value is 'C:\'.
MSI (c)  []: PROPERTY CHANGE: Adding INSTALLDIR property. Its value is 'C:\'.
MSI (c)  []: PROPERTY CHANGE: Adding _ property. Its value is 'C:\runtime-development\'.
MSI (c)  []: PROPERTY CHANGE: Adding _usr property. Its value is 'C:\runtime-development\usr\'.
MSI (c)  []: PROPERTY CHANGE: Adding _usr_bin property. Its value is 'C:\runtime-development\usr\bin\'.
MSI (c)  []: Target path resolution complete. Dumping Directory table...
MSI (c)  []: Note: target paths subject to change (via custom actions or browsing)
MSI (c)  []: Dir (target): Key: TARGETDIR	, Object: C:\
MSI (c)  []: Dir (target): Key: INSTALLDIR	, Object: C:\
MSI (c)  []: Dir (target): Key: _	, Object: C:\runtime-development\
MSI (c)  []: Dir (target): Key: _usr	, Object: C:\runtime-development\usr\
MSI (c)  []: Dir (target): Key: _usr_bin	, Object: C:\runtime-development\usr\bin\
MSI (c)  []: PROPERTY CHANGE: Adding INSTALLLEVEL property. Its value is '1'.
MSI (c)  []: Note: 1: 2205 2:  3: MsiAssembly 
MSI (c)  []: Note: 1: 2228 2:  3: MsiAssembly 4:  SELECT `MsiAssembly`.`Attributes`, `MsiAssembly`.`File_Application`, `MsiAssembly`.`File_Manifest`,  `Component`.`KeyPath` FROM `MsiAssembly`, `Component` WHERE  `MsiAssembly`.`Component_` = `Component`.`Component` AND `MsiAssembly`.`Component_` = ? 
Action ended  CostFinalize. Return value 1.
MSI (c)  []: Skipping action: SetINSTALLDIR (condition is false)
```

**Reply by tristanlabelle** [reply]
This seems to be due to the interplay between three references to `INSTALLDIR`:
```
    <Directory Id="INSTALLDIR">
      ...
    </Directory>
    <SetDirectory Id="INSTALLDIR" Value="[ProgramFiles64Folder]swift" Condition="NOT INSTALLDIR" />
    <UI>
      <ui:WixUI Id="WixUI_InstallDir" />
      <Publish Dialog="WelcomeDlg" Control="Next" Event="NewDialog" Value="InstallDirDlg" Order="2" />
      <Publish Dialog="InstallDirDlg" Control="Back" Event="NewDialog" Value="WelcomeDlg" Order="2" />
    </UI>
    <Property Id="WIXUI_INSTALLDIR" Value="INSTALLDIR"></Property>
```
If there's a simpler way to express that the installation directory should default to `%ProgramFiles%\swift` but be configurable in the UI, I'd love to try it out!

**Reply by tristanlabelle** [reply]
Ok, just setting `<Directory Id="INSTALLDIR" Name="[ProgramFiles64Folder]swift">` and removing `<SetDirectory>` seems to do the trick!

**Reply by robmen** [reply]
That is basically the solution, but it looks so weird. How about this instead?

```
<StandardDirectory Id='ProgramFiles64Folder'>
   <Directory Id='INSTALLDIR' Name='swift'>
       ...
```

Declarative is always better than procedural. `SetDirectory` is a tiny bit of procedural.

**Reply by tristanlabelle** [reply]
Ah! Thanks for your help! My previous solution also did not work, I was erroneously using a stale build.

That said, the `<StandardDirectory>` does not work for `Id="WindowsVolume"`, which we use for other MSIs, as this is not a recognized enumerant. What would be the equivalent formulation? Do we have to use `<SetDirectory>` in that case?

**Reply by tristanlabelle** [reply]
I'm not sure if this is what I should be looking at, but comparing this array:

https://github.com/wixtoolset/wix4/blob/f5bda6af2abbcf7e4792d299b86bfcec713fc6c2/src/api/wix/WixToolset.Data/WindowsInstaller/WindowsInstallerStandard.cs#L367

and this reference:

https://learn.microsoft.com/en-us/windows/win32/msi/property-reference#system-folder-properties

`WindowsVolume` is the only missing entry

**Reply by robmen** [reply]
`WindowsVolume` isn't a standard directory; it's a property (confusing as that is, see: #7233 and #7314). I would use `SetProperty` before `CostInitialize` since `SetDirectory` can force a recosting which can be slow.

**Reply by tristanlabelle** [reply]
Thank you! I got it to work using those links!
